client/blocks/payment-request/hooks.js

```diff
@@ -82,15 +102,17 @@ export const useShippingAddressUpdateHandler = (
 	paymentRequestType
 ) => {
 	useEffect( () => {
-		// Need to use `?.` here in case paymentRequest is null.
-		const shippingAddressUpdateHandler = paymentRequest?.on(
-			'shippingaddresschange',
-			shippingAddressChangeHandler( paymentRequestType )
-		);
+		let handler = noop;
+
+		if ( paymentRequest ) {
+			handler = paymentRequest.on(
+				'shippingaddresschange',
+				shippingAddressChangeHandler( paymentRequestType )
+			);
+		}
 
 		return () => {
-			// Need to use `?.` here in case shippingAddressUpdateHandler is null.
-			shippingAddressUpdateHandler?.removeAllListeners();
+			handler.removeEventListener( 'shippingaddresschange' );
 		};
 	}, [ paymentRequest, paymentRequestType ] );
```

Comment:
I'd still _prefer_ to not use a `let` binding at all. It's nitpick-y in small functions like these, but I'd still prefer to use `const` and optional chaining instead of `let` and `if`:

```js
useEffect( () => {
    const handler = paymentRequest?.on(
        'shippingaddresschange',
        shippingAddressChangeHandler( paymentRequestType )
    );

    return () => {
        handler?.removeEventLIstener( 'shippingaddresschange' );
    };
}, [ paymentRequest, paymentRequestType ] );
```

There's no need to reason about whether `paymentRequest` is null or not — it's automatically handled by the optional chaining so the additional overhead of parsing the `if` statement is removed.

That said, I don't think it's a required change to merge. Just a strong preference on my part 🙂.

Reply:
Slack discussion started in p1624477548460700-slack-C01BZUL57SQ

Reply:
Sounds good to me – done in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1603/commits/d82f0e2c398e218385fab76a3ea26383634dfd01 👍